[WIP] Allow compilation to in memory stream, part 1: Fsc output

enricosada opened the pull request:
this pr is the first of a multuple set to allow in memory compilation and output

see fsharp/FSharp.Compiler.Service#266
user voice: https://fslang.uservoice.com/forums/245727-f-language/suggestions/6717137-allow-for-completely-in-memory-compilation

the refactoring plan is to start with the lower level, convert paths to stream factories and bubble up these arguments up to fsc main.
There we can build a better api for in memory compilation ( really nice for testing or compiler service i think)

current fsc main:
```fsharp
    main0(argv,bannerAlreadyPrinted,exiter, errorLoggerProvider, disposables)
    |> main1
    |> main2
    |> main2b
    |> main2c
    |> main3 
    |> main4
```

inside `main0` the command line argument are parsed and outfile/pdbfile path is set and passed to all other funcions ( `main1` to `main4` )

this pr add output/pdb/mdb to stream or file. 
The output of assembly directly to memorystream works, pdb/mdb/signed require disk io with temp file.

the api to specify where send out/pdb/mdb is

```fsharp
type EmitTo =
    | EmittedFile of string
    | EmittedStream of Stream
``` 

with error and deferred creation

```fsharp
type EmitStreamProvider = Lazy<Choice<EmitTo, Diagnostic>>
type Diagnostic = string
```

1. `Choice<EmitTo, Diagnostic>`:
  -  to use success/fail pattern instead of exception based control flow
  - we can refactor (later) the control flow as computation expression  
2. `Lazy`:
  -  deferred creation
  - once
  - we can ask ( `Lazy.IsValueCreated` ) if the factory is called, and easy cleanup the files on error
3. both `EmittedFile of string` and `EmittedStream of Stream` , not only `Stream`
  - performance, useless continue to copy stream (build+sign, pdb) if i need a real file
  - current implementation (pdb/sign) require file on disk, useless hide that complexity. if we later improve pdb/sign, this api is ok 
  - easy to consume api

As side note, roslyn do pretty much the same with [EmitStreamProvider](http://source.roslyn.codeplex.com/#Microsoft.CodeAnalysis/Compilation.EmitStreamProvider.cs,0a7a5f6ff635682d,references), so we can wrap later when a bridge is ready ( working on a partial bridge :smile: )

i need a feedback about naming, i think is weak (for example `EmittedFile` -> `EmitToFile`? dunno)

dsyme commented:
@enricosada To clarify - this looks like an FCS feature, but you'd like the core implementation diff that enables the feature to be integrated into Microsoft/visualfsharp to keep maximal code alignment?  If so I am assuming that you are confident that the feature is non-intrusive and non-risky.  Also, just to check, do you think it will commute well with the CoreCLR work?

@KevinRansom If so I'm ok with the approach of commiting the core of FCS implementation feature work to Microsoft/visualfsharp first, then integrating through to FCS.  If you like that approach, then we should probably bring over the core implementation diff of a couple of other FCS features like in-memory cross-project references too.  That will continue to put us in a really good position to fully align in the future (e.g. having the ability to build compatible FSharp.Compiler.Service DLLs out of Microsoft/visualfsharp).

@enricosada If @KevinRansom doesn't feel able to pull in the next couple of days for stability reasons, you can submit first to FCS.  Make sure the diff is absolutely minimal.

Thanks 
Don

dsyme commented on src/absil/ilwrite.fs:
```diff
@@ -222,7 +222,6 @@ type PdbData =
 //---------------------------------------------------------------------
 
 let WritePdbInfo fixupOverlappingSequencePoints showTimes f fpdb info = 
-    (try FileSystem.FileDelete fpdb with _ -> ())
```
What's this removal for?

enricosada replied:
removed because useless, the file is now removed in `serializeToFile` (the same for the .mdb)

dsyme commented on src/absil/ilwrite.fs:
```diff
@@ -3875,7 +3874,28 @@ let writeDirectory os dict =
+
+
+// Either
+let (|Success|Failure|) = function
```
This looks needless, just use Choice1Of2 and Choice2Of2

enricosada replied:
About that, i think `Success`/`Failure` make the code easier to read instead of `Choice1Of2` / `Choice2Of2` .

i can declare a type

```fsharp
type Result<'S, 'F> = Success of 'S | Failure of 'F
```

but Choice is the same, with a little help of active pattern. But maybe the custom DU is better?

I really like to refactor to change the code from exception workflow -> Success/Failure -> computation expression because is times easier to read and is easier to use

dsyme replied:
Please just keep Choice1Of2/Choice2Of2 and remove the active pattern.  That's in keeping with enough of the style of the rest of the compiler and this PR isn't the right place to change the style. THanks

dsyme commented on src/absil/ilwrite.fs:
```diff
@@ -4025,17 +4037,18 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
           let globalpointerCodeChunk,next = chunk (if isItanium then 0x8 else 0x0) next
           
-          let debugDirectoryChunk,next = chunk (if pdbfile = None then 0x0 else sizeof_IMAGE_DEBUG_DIRECTORY) next
+          let debugDirectoryChunk,next = chunk (if pdbfile = false then 0x0 else sizeof_IMAGE_DEBUG_DIRECTORY) next
```
use ``not pdbfile``

dsyme replied:
Please make this requested change, thanks

enricosada replied:
@dsyme , sry, my bad, when there is no comment, is ok for me to change it. I'll always write reply next time

dsyme commented on src/absil/ilwrite.fs:
```diff
@@ -4025,17 +4037,18 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
+          // this in after we've written the binary. We approximate the size to MAXPATH according 
           // to what PDB writers seem to require and leave extra space just in case... 
           let debugDataJustInCase = 40
+          let MAXPATH = 260
```
Where's this constant coming from?

enricosada replied:
this is windows [MAX_PATH](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396#maxpath), the maximum length of a file path

dsyme commented on src/absil/ilwrite.fs:
```diff
@@ -4462,105 +4474,146 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
-                dprintf "iddMinorVersion = %ld\n" idd.iddMinorVersion
-                dprintf "iddType = %ld\n" idd.iddType
-                dprintf "iddData = (%A) = %s\n" idd.iddData (System.Text.Encoding.UTF8.GetString idd.iddData) *)
+            // write the IMAGE_DEBUG_DIRECTORY 
```
Please minimize this diff by replacing the try-with with a bing-end block and keeping the indentation the same

enricosada replied:
nice trick!

dsyme commented on src/absil/ilwrite.fs:
```diff
@@ -4025,17 +4037,18 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
-                                            + System.Text.Encoding.Unicode.GetByteCount(f) // See bug 748444
+                                | false -> 0x0 
+                                | true -> (24 
+                                            + MAXPATH
```
To get into Microsoft/visualfsharp you'll need to have written the code so that every point we can see that it's 100% "safe" - i.e. there is absolutely zero risk of behaviour change.  So for this kind of thing you will need need to take the max between MAXPATH and GetByteCount, or leave the behaviour exactly the same as before when using a on-disk file as output.

enricosada replied:
previous: size variable, but the upper bound is MAX_PATH
now: directly MAX_PATH 

is pretty much the same, but i dont need to know the pdb file name (because help a lot the refactoring)

dsyme replied:
How do we know the upper bound was MAX_PATH before?  I don't see anything in this bit of code limiting to that?

dsyme commented on src/absil/ilwrite.fs:
```diff
@@ -4116,7 +4129,6 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
               match p with 
               | None -> () 
               | Some pExpected -> 
-                  os.Flush(); 
```
What's this removal and why is it safe?

enricosada replied:
`os.BaseStream` is a `MemoryStream` now instead of a `FileStream` so `Flush` is useless. 
[Because any data written to a MemoryStream object is written into RAM, this method is redundant](https://msdn.microsoft.com/en-us/library/system.io.memorystream.flush(v=vs.110).aspx)
I added `os.Flush` after the `gen` call

But `os` is a BinaryStream so maybe we need to leave that or add `AutoFlush = true` ( see [AutoFlush](https://msdn.microsoft.com/en-us/library/system.io.streamwriter.autoflush%28v=vs.110%29.aspx?f=255&MSPPError=-2147217396) ). I'll readd this `os.Flush`, less diff. Good catch :smile:

dsyme commented on src/absil/ilwrite.fs:
```diff
@@ -4462,105 +4474,146 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
-          os.Close()
-          
-          try 
-              FileSystemUtilites.setExecutablePermission outfile
```
The call to setExecutablePermission was surrounded by a try-with - has that been maintained in the code?  Again we need to see that there is zero chance of behaviour change for on-disk file emit. thanks

enricosada replied:
Sure, i'll try_catch all `FileSystemUtilites.setExecutablePermission` invocations

enricosada replied:
done, is below (because need to be done only for EmittedFile)

```fsharp
path |> trySetExecutablePermission
```

dsyme commented on src/absil/ilwrite.fs:
```diff
@@ -4462,105 +4474,146 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
     mappings
 
 
+let private optionOfLazy (l: Lazy<'T>) =
```
Why is this needed?  Observing the is-this-evaluated state of lazy values is a bit of an anti-pattern and definitely best avoided

enricosada replied:
i'll defend the use of Lazy in a separate comment because i think is usefull and is a strong contract.
But i can inline (as inline refactoring) this function so less confusion ( + a comment where used about the clean used files behaviour )

dsyme commented on src/fsharp/fsc.fs:
```diff
@@ -1696,15 +1696,27 @@ module FileWriter =
+                let toFile path = lazy(ILBinaryWriter.EmitTo.EmittedFile path |> Choice1Of2)
+                let toFileOpt = Option.map toFile
+                
+                let pdbfileOpt = if not(runningOnMono) then pdbfile |> toFileOpt else None
```
No parens please after ``not``

dsyme commented on src/absil/ilwrite.fsi:
```diff
@@ -15,17 +15,22 @@ type ILStrongNameSigner =
     static member OpenKeyPairFile: string -> ILStrongNameSigner
     static member OpenKeyContainer: string -> ILStrongNameSigner
 
+type EmitStreamProvider = Lazy<Choice<EmitTo, Diagnostic>>
```
I find using ``Lazy`` here is a bit awkward - in general it's just rarely a good type to expose in APIs.  I think I'd prefer just 

    [<RequireQualifiedAccess>]
    type EmitTo = 
        | File of string
        | Stream of (unit -> Stream)

    type options = 
        { .. pdbfile : EmitTo option ...  }    

Internally to ilwrite.fs you can convert to a register that amortizes the evaluation in the EmitTo case.  If tracking diagnostics is a problem then I'm OK with you using 

        | Stream of (unit -> Choice<Stream, exn>)

or

        | Stream of (unit -> Choice<Stream, Diagnostic>)

Could you see if that works?

enricosada replied:
I'd like to encode contract of  the behaviour in the type system.

for me `(unit -> Stream)` is a problem in the contract api for the Stream:
- can be consumed multiple times, more difficult to call the api (need to dispose multiple streams)
- each call can return a different Stream
- i dont know if it'is already called ( i can track, but it'is easier to already know with `Lazy.IsValueCreated` ).

`Lazy` here is about "do that deferred, but exactly once", i can encode this requirement with a type and a DU, but lazy as type already exists.

About the other improvement, the type

```fsharp
| File of string
| Stream of (unit -> Choice<Stream, Diagnostic>)
```

seem better because i dont expected the string filename to return an error.
An api where the type system enforce that an error is not possible is better. :+1:  I'll use your version

Another way of postpone the problem is to check that at higher level and pass *internally* a non lazy 

```
type EmitTo = File of string | Stream of Stream
```
to all functions below like `writeBinaryAndReportMappings`

- pro: easier code
- pro: the runtime error opening the stream ( `unit -> Choice<Stream,Diagnostic>` ) stop fsc **before** compiling
- cons: the factory ( `unit -> Choice<Stream,Diagnostic>` is called always, but not deferred. The stream is open the whole compilation (not always a bad thing)

@dsyme maybe i can open a discussion in FCS about the api? This is an internal implementation detail, but the problem exists if a public api is exposed, so maybe is better to use same solution.

dsyme replied:
@enricosada Using Lazy in this way is not normal in the compiler.  Code that relies on observing whether lazy values have been created or not is just too subtle.  

I'd rather you somehow made the has-this-stream-been-created information more explicit somehow, rather than relying on Lazy.IsValueCreated.  

thanks

enricosada replied:
@dsyme ok about remove the Lazy, only question is 
```fsharp
EmittedStream of unit -> Choice<Stream,Diagnostic>
``` 
or directly
```
EmittedStream of Stream
```

i prefer the second, because if the user need a file, should use `EmittedFile`, otherwise should pass directly the `Stream`. 
If the user passa a `FileStream`/`MemoryMappedFile` with `EmittedStream` he should manage the dispose/cleanup (not needed for `MemoryStream` )
Code easier to follow, no need for lazy or deferred creation.

Note that fsc after refactoring use internally  `EmittedFile` not `EmittedStream`

dsyme replied:
Yes, I think I prefer the second.  I'd imagine the final exposed FCS API will just accept a Stream anyway, won't it, and the ultimate caller of FCS will own closing it.

enricosada replied:
let's use `EmittedStream of Stream` so. 
We can refactor later in fcs if needed, it's not too bad (less diff here too)

dsyme commented on src/absil/ilwrite.fs:
```diff
@@ -3875,7 +3874,28 @@ let writeDirectory os dict =
+
+let private tryDeleteFile path = 
+    try
+        if FileSystem.SafeExists path
```
``then`` at end of line please

    if ABC then
        DEF

dsyme commented on src/absil/ilwrite.fs:
```diff
@@ -3916,15 +3936,7 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
-    let  pdbData,debugDirectoryChunk,debugDataChunk,textV2P,mappings =
-        try 
-      
+    let gen os (pdbfile: bool) =
```
Is it possible to avoid having a new inner function here?  It is only called in one place.   Why not put the prologue that computes "os" and "pdbfile" in the portion directly above this?  I'm looking to minimize the diff and keep thee current overall function structure the same

enricosada replied:
ok , the `gen` function is like that after several refactoring attempts.
I'd have liked to move this function outside , but the diff was too big, so I left it like that.

I always use inner function because mvoe value binding later, so i can easy see values captured inside big functions like this one.

I will do as you said, less diff.

dsyme commented on src/absil/ilwrite.fs:
```diff
@@ -4462,105 +4474,146 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
+            writeInt32 os2 debugDataChunk.addr  // IMAGE_DEBUG_DIRECTORY.AddressOfRawData 
+            writeInt32 os2 (textV2P debugDataChunk.addr)// IMAGE_DEBUG_DIRECTORY.PointerToRawData 
+
+            (* dprintf "idd.iddCharacteristics = %ld\n" idd.iddCharacteristics
```
You can delete this stuff

dsyme commented on src/absil/ilwrite.fs:
```diff
@@ -4462,105 +4474,146 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
+    let serializeToPdb = serializeToFile ".pdb"
+    let serializeToMdb = serializeToFile ".mdb"
+
+    let emitOrFail (e: EmitStreamProvider) = 
```
Please pull this out to be a top-level function since it is independent

dsyme commented on src/absil/ilwrite.fs:
```diff
@@ -4462,105 +4474,146 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
-    //Finished writing and signing the binary and debug info...
+        reportTime showTimes "Signed Image"
+
+    let serializeToFile extension f emitter =
```
This looks independent of context (no free variables) - please pull it out to be a top level function

dsyme commented on src/absil/ilwrite.fs:
```diff
@@ -4462,105 +4474,146 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
+        reportTime showTimes "Finalize PDB"
+
+    /// Sign the binary.  No further changes to binary allowed past this point!
+    let signTo filePath (s: ILStrongNameSigner) =
```
This function looks independent of context (no free variables) - please pull it out to be a top-level function with a /// comment explaining what it does

dsyme commented on src/absil/ilwrite.fs:
```diff
@@ -4462,105 +4474,146 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
+
+    let writeDumpDebugInfo pdbData dumpTo =
+        reportTime showTimes "Dump Debug Info"
+        let dump stream = DumpDebugInfo stream pdbData
```
Please avoid little nested helper functions like this - inlining the call to DumpDebugInfo in two places is clearer to read than a nested inner function just called ``dump``.

dsyme commented on src/absil/ilwrite.fs:
```diff
@@ -4462,105 +4474,146 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
-    if dumpDebugInfo then 
-        DumpDebugInfo outfile pdbData
+
+    let writeDumpDebugInfo pdbData dumpTo =
```
Perhaps the content of this should just be combined with DumpDebugInfo

dsyme commented:
Added a bunch of comments - thanks!  It's great to have this underway - FCS can really benefit from this feature.

enricosada commented:
@dsyme really thx for feedback, i'll fix issues.

I wrote this internal refactoring as less intrusive as possibile, with minimal diff in mind.
I sent the pr there because i'd like a smooth integration with visualfsharp, instead of the difficult merge of multiple commit later when the whole in-memory compilation is ready in fcs.
I really like to minimize merge efforts, it'is wasted time.

I think after this one is merged in visualfsharp, fcs can pull and we can deliver the in-memory compilation without waiting visualfsharp if needed (i'd like to help too).

KevinRansom commented:
Don, Enrico, mark the pr's to align with [fcs-align] I will prioritise merging them. submit the PRs to master branch.  Please don't break any public API's since the work will go into VS 2015 : update 2.

enricosada commented:
@dsyme done, following your review. 
Now both the diff and implementation is better (less nested), thx

It's open only the `EmittedStream` api, we can wait for a feedback from compiler service contributors i think ( i really like `EmittedStream of Stream` )

dsyme commented:
Great thanks, I'll take another look

dsyme commented on src/absil/ilwrite.fs:
```diff
@@ -4025,17 +4074,18 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
-                                            + System.Text.Encoding.Unicode.GetByteCount(f) // See bug 748444
+                                | false -> 0x0 
+                                | true -> (24 
+                                            + MAX_PATH
```
I think I still want this to preserve the old behaviour in the case that we have a named file.  I just don't want us to have any chance of a regression here where we do something wrong with long file names.

enricosada replied:
i'll try to pass the pdb filename as string option, a boolean was easier for me but i think i can do it with string option too.

dsyme replied:
thanks :)

dsyme commented:
A couple of comments added.  Nice work - the diff is now much cleaner!!

enricosada commented:
ok, is try to put bak pdb path and use `Stream of Stream`, thx don

enricosada commented:
done, less diff, revert to use pdb path instead of MAX_PATH, and cleanups

I'd like to test this code more, so wait a bit before merge ( review is ok :eyes: )

dsyme commented:
Fantastic! Will take another glance over the code a bit later

enricosada commented:
ok done testing:
- to stream of pdb/out/debuginfo
- to file of pdb/out/debuginfo
- `--times` flag
- `--test:DumpDebugInfo`

for me is ready to go. 
I'll add unit tests for EmitTo.Stream after the next refactoring (move arguments to `main1`)

dsyme commented:
OK, this looks good to go to me

dsyme commented:
(BTW I think I'd expect any further work on the compile-to-stream feature to go to fsharp/FSharp.Compiler.Service after this has been integrated - otherwise it should have been included here?)

enricosada commented:
I think remains to move parameters up per in the call stack ( `main1` ). I'll add another commit to this pr now that the to stream implementation is done

Inviato da Outlook




On Mon, Nov 16, 2015 at 1:37 PM -0800, "Don Syme" <notifications@github.com> wrote:
(BTW I think I'd expect any further work on the compile-to-stream feature to go to fsharp/FSharp.Compiler.Service after this has been integrated - otherwise it should have been included here?)

---
Reply to this email directly or view it on GitHub:
https://github.com/Microsoft/visualfsharp/pull/715#issuecomment-157178278

enricosada commented on src/fsharp/fsc.fs:
```diff
@@ -1840,30 +1861,38 @@ let main1(tcGlobals, tcImports: TcImports, frameworkTcImports, generatedCcu, typ
       ReportTime tcConfig ("Write XML document signatures")
-      if tcConfig.xmlDocOutputFile.IsSome then 
+      match xmlDocOutputOpt with
+      | Some(_) ->
```
converted from `if` to `match` because i need type inference, otherwise i must specify method signature for `xmlDocOutputOpt` param (more invasive diff)

enricosada commented:
ok, i need to do more testing but that should be ok to review at least /cc @dsyme. I'll do more testing asap

I changed only internal functions inside `fsc.fs`, moving all binding one step up in the call stack (`main4` -> `main3` -> .. `main0` ) so the previous review of the others files is ok.

i moved some string path -> EmitTo where possible (like xmldoc)

I cannot move the `outfile` string path upper in the call stack because i need to change `TastPickle.pickleObjWithDanglingCcus` signature inside `EncodeOptimizationData`, but this pr is already big. i can continue later with another.

btw a note, i think we need to change some members of `TcConfig`, like `TcConfig.outputFile` from `string option` to `EmitTo`, so it's easier to propagate the change from `path` to `path or stream`. Typed refactoring is really easier.

After that, we can use a wrapper of `main1` as api in the compiler, so a clean api but the same internal code of the compiler, that should help compiler service (or internally the vs extensions)

I know it's seem useless now, but i really think this refactoring can help with the coreclr because now the code is easier to change (add portable pdb output as another parameter for example)

Each step `main0` -> `main1` -> `main2` -> ... remove arguments, instead of create new ones or local bindings, so it's easier to understand.

dsyme commented on src/fsharp/fsc.fs:
```diff
@@ -1887,18 +1918,29 @@ let main2(Args(tcConfig, tcImports, frameworkTcImports: TcImports, tcGlobals, er
-            let sigDataFileName = (Filename.chopExtension outfile)+".fsdata"
-            File.WriteAllBytes(sigDataFileName,bytes)
+            //TODO instead of bytes array we can directly write to stream, less memory pressure
+            bytes |> emitAllBytes fsDataP
```
FWIW it looks to me like useSignatureDataFile is always false and this is dead code, as is anything to do with fsData files.

enricosada commented:
i'll rebase to fix emrge conflict

KevinRansom commented:
@enricosada Enrico, in your opinion is this ready to merge?

enricosada commented:
Yes, after rebase.

enricosada commented:
I'll do that soon, this week

dsyme commented:
Marked WIP until rebase is done (please remove once it is done)

dsyme commented:
@enricosada Can you bring this up to date with CI tests passing?  It would be good to get it merged. Thanks

enricosada commented:
@dsyme i'll wait the merge of `coreclr` branch inside `master`. 
I dont want to make the merge more difficult for @KevinRansom because of pdb work

KevinRansom commented:
@enricosada   coreclr is now merged into master and pretty stable, would you like to take care of this?

Thanks

Kevin

KevinRansom commented:
@enricosada,

Hey ... do you want to finish this off or shall I close this PR?

Thanks

Kevin

enricosada commented:
i'll close now and reopen a new pr after a rebase, too much diff atm
